src/commands/cp.rs

```diff
@@ -55,26 +123,277 @@ pub fn cp(args: CommandArgs) -> Result<OutputStream, ShellError> {
+                                e.to_string(),
+                                e.to_string(),
+                                name_span,
+                            ));
+                        }
+                        Ok(o) => o,
+                    };
+
+                    for (ref src, ref dst) in sources.translate(|(src, loc)| {
+                        let mut final_path = destination.clone();
+                        let path = dunce::canonicalize(&src).unwrap();
+
+                        let mut comps: Vec<_> = path
+                            .components()
+                            .map(|fragment| fragment.as_os_str())
+                            .rev()
+                            .take(1 + loc)
+                            .collect();
+
+                        comps.reverse();
+
+                        for fragment in comps.iter() {
+                            final_path.push(fragment);
+                        }
+
+                        (PathBuf::from(&src), PathBuf::from(final_path))
+                    }) {
+                        if src.is_dir() {
+                            if !dst.exists() {
+                                match std::fs::create_dir_all(dst) {
+                                    Err(e) => {
+                                        return Err(ShellError::labeled_error(
+                                            e.to_string(),
+                                            e.to_string(),
+                                            name_span,
+                                        ));
+                                    }
+                                    Ok(o) => o,
+                                };
+                            }
+                        }
+
+                        if src.is_file() {
+                            match std::fs::copy(src, dst) {
+                                Err(e) => {
+                                    return Err(ShellError::labeled_error(
+                                        e.to_string(),
+                                        e.to_string(),
+                                        name_span,
+                                    ));
+                                }
+                                Ok(o) => o,
+                            };
+                        }
+                    }
+                } else {
+                    destination.push(val.file_name().unwrap());
+
+                    match std::fs::create_dir_all(&destination) {
+                        Err(e) => {
+                            return Err(ShellError::labeled_error(
+                                e.to_string(),
+                                e.to_string(),
+                                name_span,
+                            ));
+                        }
+                        Ok(o) => o,
+                    };
+
+                    for (ref src, ref dst) in sources.translate(|(src, loc)| {
```

Comment:
Can we do these as separate steps? It'd just be easier to read the code if the translate block didn't look like part of the `for`